File: api/ubicquia.py
# Vercel Python Serverless Function: /api/ubicquia
# Replica el flujo del script Python que descarga todas las columnas.
# Usa variables de entorno:
# UBICQUIA_<PANEL>_CLIENT_ID
# UBICQUIA_<PANEL>_CLIENT_SECRET
# Opcional: UBICQUIA_ACCESS_CODE o ACCESS_CODE o DOWNLOAD_CODE

from http.server import BaseHTTPRequestHandler
import os
import json
import requests
from urllib.parse import quote

TOKEN_URL = "https://auth.ubihub.ubicquia.com/auth/realms/ubivu-prd/protocol/openid-connect/token"
DTM_METRIX_URL = "https://api.ubicquia.com/api/ubigrid/transformer/metrix/list"
NOTIFICATIONS_URL = "https://api.ubicquia.com/api/v2/notification-nodes"
PER_PAGE = 20000
MAX_PAGES = 100
BACKEND_VERSION = "python-backend-v1-exact-reference-flow"


def response(handler, status, payload):
    body = json.dumps(payload, ensure_ascii=False, default=str).encode("utf-8")
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json; charset=utf-8")
    handler.send_header("Cache-Control", "no-store")
    handler.send_header("Access-Control-Allow-Origin", "*")
    handler.send_header("Access-Control-Allow-Headers", "Content-Type, x-access-code")
    handler.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
    handler.end_headers()
    handler.wfile.write(body)


def get_body(handler):
    length = int(handler.headers.get("Content-Length", "0") or "0")
    raw = handler.rfile.read(length) if length else b"{}"
    return json.loads(raw.decode("utf-8") or "{}")


def check_access(handler):
    expected = os.environ.get("UBICQUIA_ACCESS_CODE") or os.environ.get("ACCESS_CODE") or os.environ.get("DOWNLOAD_CODE")
    if not expected:
        return True
    got = handler.headers.get("x-access-code")
    return got == expected


def get_credentials(panel):
    key = str(panel or "646703").strip()
    return key, os.environ.get(f"UBICQUIA_{key}_CLIENT_ID"), os.environ.get(f"UBICQUIA_{key}_CLIENT_SECRET")


def get_access_token(client_id, client_secret):
    auth_data = {
        "grant_type": "client_credentials",
        "scope": "openid",
        "client_id": client_id,
        "client_secret": client_secret,
    }
    headers = {"Content-Type": "application/x-www-form-urlencoded"}
    r = requests.post(TOKEN_URL, data=auth_data, headers=headers, timeout=60)
    r.raise_for_status()
    token = r.json().get("access_token")
    if not token:
        raise RuntimeError("Auth response did not include access_token")
    return token


def fetch_api_data_paginated(url, headers, data):
    all_data = []
    page = 1
    while page <= MAX_PAGES:
        payload = dict(data)
        payload["page"] = page
        payload["per_page"] = str(PER_PAGE)
        r = requests.post(url, headers=headers, json=payload, timeout=90)
        r.raise_for_status()
        page_data = r.json().get("data", [])
        if not page_data:
            break
        all_data.extend(page_data)
        page += 1
    return all_data


def fetch_notifications_paginated(notification_type, start_date, end_date, access_token, imei, subpanel_id):
    headers_api = {
        "Authorization": f"Bearer {access_token}",
        "Content-Type": "application/json",
        "current-subpanel-id": str(subpanel_id),
    }
    all_notifications = []
    page = 1
    while page <= MAX_PAGES:
        url = (
            f"{NOTIFICATIONS_URL}?type=transformers"
            f"&start_date={quote(start_date)}&end_date={quote(end_date)}"
            f"&notification_type={quote(notification_type)}&page={page}&per_page={PER_PAGE}"
        )
        r = requests.get(url, headers=headers_api, timeout=90)
        r.raise_for_status()
        page_data = r.json().get("data", {}).get("nodes", [])
        if not page_data:
            break
        all_notifications.extend(page_data)
        page += 1
    return [n for n in all_notifications if str(n.get("dev_eui")) == str(imei)]


def safely_parse_json(x):
    if x is None:
        return {}
    if isinstance(x, dict):
        return x
    try:
        return json.loads(x)
    except Exception:
        try:
            return json.loads(str(x).replace("''", '"').replace("“", '"').replace("”", '"'))
        except Exception:
            return {}


def add_msg_fields(rows):
    out = []
    for r in rows or []:
        item = dict(r)
        parsed = safely_parse_json(item.get("jsonData"))
        item["MsgStr"] = parsed.get("msgStr")
        item["MsgType"] = parsed.get("msgType")
        out.append(item)
    return out


def filter_power_loss(rows):
    out = []
    for r in add_msg_fields(rows):
        if r.get("MsgStr") in ["AlertPowerLoss", "AlertPowerLoss2"] or r.get("alertvalue") == "Loss":
            out.append(r)
    return out


def filter_power_restored(rows):
    out = []
    for r in add_msg_fields(rows):
        if r.get("MsgStr") == "AlertPowerRestored" or r.get("alertvalue") == "Restored":
            out.append(r)
    return out


class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        response(self, 200, {"ok": True})

    def do_POST(self):
        try:
            if not check_access(self):
                response(self, 401, {"error": "Invalid access code"})
                return

            body = get_body(self)
            action = body.get("action", "all")
            if action != "all":
                response(self, 400, {"error": "Only action=all is supported"})
                return

            imei = str(body.get("imei", "")).strip()
            start = body.get("start")
            end = body.get("end")
            subpanel_id = str(body.get("subpanel_id") or "0")
            panel = body.get("panel") or "646703"

            if not imei or not start or not end:
                response(self, 400, {"error": "Missing imei, start or end"})
                return

            panel_key, client_id, client_secret = get_credentials(panel)
            if not client_id or not client_secret:
                response(self, 500, {"error": f"Missing credentials for panel {panel_key}"})
                return

            token = get_access_token(client_id, client_secret)
            headers_api = {
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json",
                "current-subpanel-id": subpanel_id,
            }

            post_data = {
                "start_datetime": start,
                "end_datetime": end,
                "imei": imei,
                "serialNumber": imei,
                "type": "voltage",
                "page": "1",
                "per_page": str(PER_PAGE),
            }

            metrix = fetch_api_data_paginated(DTM_METRIX_URL, headers_api, post_data)
            sag = fetch_notifications_paginated("AlertAggregatedVoltageSag", start, end, token, imei, subpanel_id)
            swell = fetch_notifications_paginated("AlertAggregatedVoltageSwell120/240/277V", start, end, token, imei, subpanel_id)
            loss_raw = fetch_notifications_paginated("AlertPowerLoss", start, end, token, imei, subpanel_id)
            restored_raw = fetch_notifications_paginated("AlertPowerRestored", start, end, token, imei, subpanel_id)

            powerloss = filter_power_loss(loss_raw)
            powerrestored = filter_power_restored(restored_raw)

            response(self, 200, {
                "data": {
                    "backendVersion": BACKEND_VERSION,
                    "metrix": metrix,
                    "sag": sag,
                    "swell": swell,
                    "powerloss": powerloss,
                    "powerrestored": powerrestored,
                    "counts": {
                        "metrix": len(metrix),
                        "sag": len(sag),
                        "swell": len(swell),
                        "powerloss": len(powerloss),
                        "powerrestored": len(powerrestored),
                    }
                }
            })
        except Exception as e:
            response(self, 500, {"error": str(e)})

File: api/ubicquia.js
// Vercel /api/ubicquia.js
// Descarga completa estilo script Python: metrix paginado + notificaciones paginadas.
// Requiere variables de entorno por panel:
// UBICQUIA_<PANEL>_CLIENT_ID
// UBICQUIA_<PANEL>_CLIENT_SECRET
// Opcional: UBICQUIA_ACCESS_CODE o ACCESS_CODE o DOWNLOAD_CODE

const TOKEN_URL = 'https://auth.ubihub.ubicquia.com/auth/realms/ubivu-prd/protocol/openid-connect/token';
const DTM_METRIX_URL = 'https://api.ubicquia.com/api/ubigrid/transformer/metrix/list';
const NOTIFICATIONS_URL = 'https://api.ubicquia.com/api/v2/notification-nodes';
const PER_PAGE = 20000;
const MAX_PAGES = 100;

function pageSignature(rows) {
  const first = rows[0] || {};
  const last = rows[rows.length - 1] || {};
  return [
    rows.length,
    first.id || first.createdAt || first.date || first.dev_eui || '',
    last.id || last.createdAt || last.date || last.dev_eui || ''
  ].join('|');
}

function getEnv(name) {
  return process.env[name];
}

function normalizePanel(panel) {
  return String(panel || '646703').trim();
}

function getCredentials(panel) {
  const key = normalizePanel(panel);
  const clientId = getEnv(`UBICQUIA_${key}_CLIENT_ID`);
  const clientSecret = getEnv(`UBICQUIA_${key}_CLIENT_SECRET`);
  return { key, clientId, clientSecret };
}

function requireAccessCode(req, res) {
  const expected = getEnv('UBICQUIA_ACCESS_CODE') || getEnv('ACCESS_CODE') || getEnv('DOWNLOAD_CODE');
  if (!expected) return true;
  const got = req.headers['x-access-code'];
  if (got !== expected) {
    res.status(401).json({ error: 'Invalid access code' });
    return false;
  }
  return true;
}

async function getAccessToken(clientId, clientSecret) {
  const params = new URLSearchParams();
  params.set('grant_type', 'client_credentials');
  params.set('scope', 'openid');
  params.set('client_id', clientId);
  params.set('client_secret', clientSecret);

  const response = await fetch(TOKEN_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: params.toString(),
  });

  const text = await response.text();
  if (!response.ok) {
    throw new Error(`Auth HTTP ${response.status}: ${text}`);
  }
  const json = JSON.parse(text);
  if (!json.access_token) throw new Error('Auth response did not include access_token');
  return json.access_token;
}

async function postJson(url, headers, body) {
  const response = await fetch(url, {
    method: 'POST',
    headers,
    body: JSON.stringify(body),
  });
  const text = await response.text();
  if (!response.ok) throw new Error(`POST ${url} HTTP ${response.status}: ${text}`);
  return text ? JSON.parse(text) : {};
}

async function getJson(url, headers) {
  const response = await fetch(url, { method: 'GET', headers });
  const text = await response.text();
  if (!response.ok) throw new Error(`GET ${url} HTTP ${response.status}: ${text}`);
  return text ? JSON.parse(text) : {};
}

async function fetchMetrixAllPages({ token, subpanelId, imei, start, end }) {
  const headers = {
    Authorization: `Bearer ${token}`,
    'Content-Type': 'application/json',
    'current-subpanel-id': String(subpanelId || '0'),
  };

  const all = [];
  const seen = new Set();
  for (let page = 1; page <= MAX_PAGES; page++) {
    const payload = {
      start_datetime: start,
      end_datetime: end,
      imei: String(imei).trim(),
      serialNumber: String(imei).trim(),
      type: 'voltage',
      page: String(page),
      per_page: String(PER_PAGE),
    };

    const json = await postJson(DTM_METRIX_URL, headers, payload);
    const rows = Array.isArray(json.data) ? json.data : [];
    if (!rows.length) break;
    const sig = pageSignature(rows);
    if (seen.has(sig)) break;
    seen.add(sig);
    all.push(...rows);
    // Importante: NO detener por rows.length < PER_PAGE.
    // El script Python de referencia sigue pidiendo páginas hasta recibir una página vacía.
  }
  return all;
}

async function fetchNotificationsAllPages({ token, subpanelId, imei, start, end, notificationType }) {
  const headers = {
    Authorization: `Bearer ${token}`,
    'Content-Type': 'application/json',
    'current-subpanel-id': String(subpanelId || '0'),
  };

  const all = [];
  const seen = new Set();
  for (let page = 1; page <= MAX_PAGES; page++) {
    const url = `${NOTIFICATIONS_URL}?type=transformers` +
      `&start_date=${encodeURIComponent(start)}` +
      `&end_date=${encodeURIComponent(end)}` +
      `&notification_type=${encodeURIComponent(notificationType)}` +
      `&page=${page}&per_page=${PER_PAGE}`;

    const json = await getJson(url, headers);
    const nodes = json && json.data && Array.isArray(json.data.nodes) ? json.data.nodes : [];
    if (!nodes.length) break;
    const sig = pageSignature(nodes);
    if (seen.has(sig)) break;
    seen.add(sig);
    all.push(...nodes.filter(n => String(n.dev_eui || '') === String(imei)));
    // Igual que el Python: seguir hasta página vacía, no detener por tamaño menor a PER_PAGE.
  }
  return all;
}

function safelyParseJson(x) {
  if (x == null) return {};
  if (typeof x === 'object') return x;
  try { return JSON.parse(x); }
  catch (_) {
    try { return JSON.parse(String(x).split("''").join('"').split('“').join('"').split('”').join('"')); }
    catch (__) { return {}; }
  }
}

function addMsgFields(rows) {
  return (rows || []).map(r => {
    const d = safelyParseJson(r.jsonData);
    return { ...r, MsgStr: d.msgStr, MsgType: d.msgType };
  });
}

function filterPowerLoss(rows) {
  return addMsgFields(rows).filter(r => ['AlertPowerLoss', 'AlertPowerLoss2'].includes(r.MsgStr) || r.alertvalue === 'Loss');
}

function filterPowerRestored(rows) {
  return addMsgFields(rows).filter(r => r.MsgStr === 'AlertPowerRestored' || r.alertvalue === 'Restored');
}

export default async function handler(req, res) {
  try {
    if (req.method !== 'POST') {
      res.status(405).json({ error: 'Method not allowed' });
      return;
    }
    if (!requireAccessCode(req, res)) return;

    const body = req.body || {};
    const action = body.action || 'all';
    const imei = String(body.imei || '').trim();
    const start = body.start;
    const end = body.end;
    const subpanelId = body.subpanel_id || '0';
    const { key, clientId, clientSecret } = getCredentials(body.panel);

    if (!imei || !start || !end) {
      res.status(400).json({ error: 'Missing imei, start or end' });
      return;
    }
    if (!clientId || !clientSecret) {
      res.status(500).json({ error: `Missing credentials for panel ${key}` });
      return;
    }

    const token = await getAccessToken(clientId, clientSecret);

    if (action !== 'all') {
      res.status(400).json({ error: 'Only action=all is supported by this endpoint' });
      return;
    }

    const [metrix, sag, swell, lossRaw, restoredRaw] = await Promise.all([
      fetchMetrixAllPages({ token, subpanelId, imei, start, end }),
      fetchNotificationsAllPages({ token, subpanelId, imei, start, end, notificationType: 'AlertAggregatedVoltageSag' }),
      fetchNotificationsAllPages({ token, subpanelId, imei, start, end, notificationType: 'AlertAggregatedVoltageSwell120/240/277V' }),
      fetchNotificationsAllPages({ token, subpanelId, imei, start, end, notificationType: 'AlertPowerLoss' }),
      fetchNotificationsAllPages({ token, subpanelId, imei, start, end, notificationType: 'AlertPowerRestored' }),
    ]);

    const powerloss = filterPowerLoss(lossRaw);
    const powerrestored = filterPowerRestored(restoredRaw);

    res.status(200).json({
      data: {
        metrix,
        sag,
        swell,
        powerloss,
        powerrestored,
        counts: {
          metrix: metrix.length,
          sag: sag.length,
          swell: swell.length,
          powerloss: powerloss.length,
          powerrestored: powerrestored.length,
        },
      },
    });
  } catch (error) {
    res.status(500).json({ error: error.message || String(error) });
  }
}
